### PR TITLE
Add CORS support and set response header 'Access-Control-Allow-Origin'

### LINKED ISSUE #4326: [dev.icinga.com #12042] API should provide CORS Header
This issue has been migrated from Redmine: https://dev.icinga.com/issues/12042

**Created by rhillmann on 2016-06-25 17:14:59 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _Backlog_
Last Update: _2016-09-07 11:39:52 +00:00 (in Redmine)_

    Backport?: Not yet backported
    Include in Changelog: 1

---

I am missing cross origin header for the API, this will help to use the API directly within javascript in the Browser.

### PR DESCRIPTION
**Pull request by N-o-X**
fixes #4326

### CODE REVIEW

**Comment by dnsmichi** [comment]
- please add your tests and outputs @N-o-X 
- header\_allow\_origin should be renamed to reflect the header option (or we use a dictionary with keys here)
- we should discuss if it is possible to set multiple origin URLs. I remember that Grafana supports that. 

https://stackoverflow.com/questions/42337877/change-grafanas-headers-access-control-allow-origin

also mentions more options, i.e. to limit the methods. Think of a JS library in browser which should just GET things.

```
Header set Access-Control-Allow-Origin    "http://mygrafana.com:5656"
Header set Access-Control-Allow-Methods   "GET, OPTIONS"
Header set Access-Control-Allow-Headers   "origin, authorization, accept"
Header set Access-Control-Allow-Credentials true
```

**Comment by N-o-X** [comment]
Tested with http://test-cors.org/

**Requirements:**
- Add `access_control_allow_origin = [ "http://www.test-cors.org" ]` to api.conf
- API User should be `root:icinga`

**Tests:**
- [GET](http://www.test-cors.org/#?client_method=GET&client_credentials=true&server_url=https%3A%2F%2Flocalhost%3A5665%2Fv1%2Fobjects%2Fhosts&server_enable=true&server_status=200&server_credentials=false&server_tabs=remote&client_headers=Authorization%3A%20Basic%20cm9vdDppY2luZ2E%3D)
- [POST](http://www.test-cors.org/#?client_method=POST&client_credentials=true&client_headers=Authorization%3A%20Basic%20cm9vdDppY2luZ2E%3D%0AAccept%3A%20application%2Fjson&server_url=https%3A%2F%2Flocalhost%3A5665%2Fv1%2Factions%2Fschedule-downtime%3Ftype%3DService%26filter%3Dservice.name%3D%3D%2522ping4%2522&server_enable=true&server_status=200&server_credentials=false&server_tabs=remote&client_postdata=%7B%20%22start_time%22%3A%201446388806%2C%20%22end_time%22%3A%201446389806%2C%20%22duration%22%3A%201000%2C%20%22author%22%3A%20%22icingaadmin%22%2C%20%22comment%22%3A%20%22IPv4%20network%20maintenance%22%20%7D)
- [PUT](http://www.test-cors.org/#?client_method=PUT&client_credentials=true&client_headers=Authorization%3A%20Basic%20cm9vdDppY2luZ2E%3D%0AAccept%3A%20application%2Fjson&server_url=https%3A%2F%2Flocalhost%3A5665%2Fv1%2Fobjects%2Fhosts%2Fexample.localdomain&server_enable=true&server_status=200&server_credentials=false&server_tabs=remote&client_postdata=%7B%20%22templates%22%3A%20%5B%20%22generic-host%22%20%5D%2C%20%22attrs%22%3A%20%7B%20%22address%22%3A%20%22192.168.1.1%22%2C%20%22check_command%22%3A%20%22hostalive%22%2C%20%22vars.os%22%20%3A%20%22Linux%22%20%7D%20%7D)
- [DELETE](http://www.test-cors.org/#?client_method=DELETE&client_credentials=true&client_headers=Authorization%3A%20Basic%20cm9vdDppY2luZ2E%3D%0AAccept%3A%20application%2Fjson&server_url=https%3A%2F%2Flocalhost%3A5665%2Fv1%2Fobjects%2Fhosts%2Fexample.localdomain%3Fcascade%3D1&server_enable=true&server_status=200&server_credentials=false&server_tabs=remote)

File: lib/remote/apilistener.ti [hunk 1]
```diff
@@ -49,6 +49,23 @@ class ApiListener : ConfigObject
...
+	[config] array(String) access_control_allow_origin {
+		default {{{ return new Array(); }}}
+	};
+	[config] bool access_control_allow_credentials
+	{
+		default {{{ return true; }}}
+	};
+	[config] String access_control_allow_headers
+	{
+		default {{{ return "Authorization"; }}}
```
Comment (dnsmichi) [inline lib/remote/apilistener.ti]: The default value should be added to the docs table :)

File: lib/remote/apilistener.ti [hunk 2]
```diff
@@ -49,6 +49,23 @@ class ApiListener : ConfigObject
...
+	};
+	[config] String access_control_allow_methods
+	{
+		default {{{ return "GET, POST, PUT, DELETE"; }}}
+	};
+
+
```
Comment (dnsmichi) [inline lib/remote/apilistener.ti]: Same as headers -> docs defaults.

File: lib/remote/httpresponse.cpp [hunk 3]
```diff
@@ -74,6 +68,10 @@ void HttpResponse::FinishHeaders(void)
 			AddHeader("Transfer-Encoding", "chunked");
 
 		AddHeader("Server", "Icinga/" + Application::GetAppVersion());
+
+		for (const String& header:m_Headers)
```
Comment (dnsmichi) [inline lib/remote/httpresponse.cpp]: Small code style - add a whitespace before `m_Headers`.

File: lib/remote/httpserverconnection.cpp [hunk 4]
```diff
@@ -171,6 +171,40 @@ void HttpServerConnection::ProcessMessageAsync(HttpRequest& request)
 
 	HttpResponse response(m_Stream, request);
 
+	Array::Ptr headerAllowOrigin = ApiListener::GetInstance()->GetAccessControlAllowOrigin();
```
Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: Better add a safety check here if we really get an ApiListener object. This prevents unwanted crashes.

File: lib/remote/httpserverconnection.cpp [hunk 5]
```diff
@@ -171,6 +171,40 @@ void HttpServerConnection::ProcessMessageAsync(HttpRequest& request)
...
+
+	if (headerAllowOrigin->GetLength() != 0)
+	{
```
Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: Parenthesis for conditions, loops, etc. should be on the same line. This applies to all the other code changes made here :)

File: lib/remote/httpserverconnection.cpp [hunk 6]
```diff
@@ -171,6 +171,40 @@ void HttpServerConnection::ProcessMessageAsync(HttpRequest& request)
...
+		String origin = request.Headers->Get("origin");
+
+		ObjectLock olock(headerAllowOrigin);
+
+		for (String allowedOrigin : headerAllowOrigin)
```
Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: Wrap it in parenthesis here, it is more than one line below.

Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: Oh, and the String should be const ref.

File: lib/remote/httpserverconnection.cpp [hunk 7]
```diff
@@ -171,6 +171,40 @@ void HttpServerConnection::ProcessMessageAsync(HttpRequest& request)
...
+		String accessControlRequestMethodHeader = request.Headers->Get("access-control-request-method");
+
+		if (!accessControlRequestMethodHeader.IsEmpty())
+		{
+			response.SetStatus(200, "OK");
+
+			response.AddHeader("Access-Control-Allow-Methods", ApiListener::GetInstance()->GetAccessControlAllowMethods());
+			response.AddHeader("Access-Control-Allow-Headers", ApiListener::GetInstance()->GetAccessControlAllowHeaders());
+
+			String msg = "<h1>Preflight OK!</h1>";
```
Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: This should be either JSON (if the `Accept: application/json` header is sent) or HTML.

File: lib/remote/httpserverconnection.cpp (hunk 4 above)
Comment (dnsmichi) [inline lib/remote/httpserverconnection.cpp]: Applies to all occurrences in the code below.

File: lib/remote/apilistener.ti (hunk 1 above)
Comment (N-o-X) [inline lib/remote/apilistener.ti]: It's already there :)

File: lib/remote/apilistener.ti (hunk 2 above)
Comment (N-o-X) [inline lib/remote/apilistener.ti]: That's already there too :)

File: lib/remote/apilistener.ti (hunk 1 above)
Comment (dnsmichi) [inline lib/remote/apilistener.ti]: Ah, my bad. Too many things yesterday, sorry :)

**Comment by dnsmichi** [comment]
I've rebased it against master.

Manual tests with curl and sending origin and requests work fine.

```
curl -u root:icinga -H 'Origin: https://localhost' -H "Access-Control-Request-Method: GET" https://localhost:5665/v1/status -v
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 5665 (#0)
* TLS 1.2 connection using TLS_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: mbmif.int.netways.de
* Server auth using Basic with user 'root'
> GET /v1/status HTTP/1.1
> Host: localhost:5665
> Authorization: Basic cm9vdDppY2luZ2E=
> User-Agent: curl/7.54.0
> Accept: */*
> Origin: https://localhost
> Access-Control-Request-Method: GET
>
< HTTP/1.1 200 OK
< Access-Control-Allow-Origin: https://localhost
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Methods: GET, POST, PUT, DELETE
< Access-Control-Allow-Headers: Authorization
< Transfer-Encoding: chunked
< Server: Icinga/v2.7.0-155-gc2ae6b180
<
* Connection #0 to host localhost left intact
Preflight OK
```

But I cannot get it to work with test-cors.org.

I've even hacked up a small sample code which connects but fails to send the authorization header.


```
<html>
<head>
<script type="text/javascript">
var createCORSRequest = function(method, url) {
  var xhr = new XMLHttpRequest();
  if ("withCredentials" in xhr) {
    // Most browsers.
    xhr.open(method, url, true);
  } else if (typeof XDomainRequest != "undefined") {
    // IE8 & IE9
    xhr = new XDomainRequest();
    xhr.open(method, url);
  } else {
    // CORS not supported.
    xhr = null;
  }
  return xhr;
};

var url = 'https://localhost:5665/v1/objects/hosts';
var method = 'GET';
var xhr = createCORSRequest(method, url);

xhr.onload = function() {
  // Success code goes here.
  alert('success')
  document.getElementById("logger-message").innerHTML="success"
};

xhr.onerror = function() {
  // Error code goes here.
  alert('error')
  document.getElementById("logger-message").innerHTML="error"
};

xhr.withCredentials = true;
xhr.setRequestHeader('Authorization', 'Basic cm9vdDppY2luZ2E=');
xhr.send();
</script>
</head>

<body>
CORS Test
<br>
Message: <div id="logger-message">

</body>
</html>
```

This always generates an error, similar to test-cors.org

@N-o-X any ideas?

**Comment by dnsmichi** [comment]
We've tested it again, and it now works. Probably I've tested it the wrong way.

```
  access_control_allow_origin = [ "https://localhost", "http://localhost", "http://www.test-cors.org" ]
```

screen shot 2017-09-20 at 13 16 07

**Comment by dnsmichi** [comment]
Thanks for the implementation 👍

**Comment by rayterrill** [comment]
I'm running into an issue trying to get CORS working on r2.8.1-1. Setting the key as defined above, but I'm not seeing the Access Control Allow Origin key being set on the browser side (or via CURL.

```
/**
 * The API listener is used for distributed monitoring setups.
 */

object ApiListener "api" {
  //accept_config = false
  //accept_commands = false

  ticket_salt = TicketSalt
  access_control_allow_origin = [ "http://www.test-cors.org" ]
}
```

I do have a ticket here: https://www.monitoring-portal.org/t/cors-configuration/945/4. Open up a new issue?

**Comment by rayterrill** [comment]
Got a 2nd confirmation this does not appear to be working. Will open an issue.

**Pull request merged**